core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala

```diff
@@ -1871,21 +1871,6 @@ private[spark] class DAGScheduler(
                     markStageAsFinished(resultStage)
                     cancelRunningIndependentStages(job, s"Job ${job.jobId} is finished.")
                     cleanupStateForJobAndIndependentStages(job)
-                    try {
-                      // killAllTaskAttempts will fail if a SchedulerBackend does not implement
-                      // killTask.
-                      logInfo(s"Job ${job.jobId} is finished. Cancelling potential speculative " +
-                        "or zombie tasks for this job")
-                      // ResultStage is only used by this job. It's safe to kill speculative or
-                      // zombie tasks in this stage.
-                      taskScheduler.killAllTaskAttempts(
```

Comment:
`cancelRunningIndependentStages` should already kill all the tasks of this stage.

Reply:
I agree:

It looks like `taskScheduler.killAllTaskAttempts` was added in https://github.com/apache/spark/pull/22771 and at that time `cancelRunningIndependentStages` did not exist. The `cancelRunningIndependentStages` method was added later in https://github.com/apache/spark/pull/27050 and is a generalization to cancel both the extra tasks from the finished stage and plus any other independent zombie stage attempts.

Reply:
It turns out that we should switch the order beween `markStageAsFinished` and `cancelRunningIndependentStages` after removing this code block. `markStageAsFinished` removes the stage from `runningStages` and `cancelRunningIndependentStages` only cleanup stages that exist in `runningStages`.

Reply:
Furthermore we can also remove `markStageAsFinished` here because `cancelRunningIndependentStages` already does that for each stage after the cleanup:

https://github.com/apache/spark/blob/14d854beeabcf872175edfd65cb6475ecb7d0ae7/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala#L2862-L2863

Reply:
See my [comment above](https://github.com/apache/spark/pull/43954/files#r1409479495), unless we have other changes, we will have to revert this.

Reply:
@mridulm Thanks. I think you're right. I'll revert the changes around here.

(I'd like to see if we have other ways to organize these functions better. It's a bit messy now.)